umbraco/Umbraco-CMS
# Migrate Get remaining seconds functionality

__This must be merged first! https://github.com/umbraco/Umbraco-CMS/pull/8225__ (and then probably netcore/netcore re-merged back into this one)

for [AB#6968](https://dev.azure.com/umbraco/243e7927-03b2-44e2-908f-d4ac7ea5daaa/_workitems/edit/6968)

* Moves and adds tests for back office cookie manager
* Removes old code from the net framework implementation that has been ported and doesn't need to be there anymore
* Changes all "UmbracoApi" strings to constants
* Moves the GetRemainingTimeoutSeconds endpoint to be on our auth controller instead of on custom middleware since there's no longer a need for that
* KeepUserLoggedIn config value should not be true by default so changed to false
* back office cookie options configured to renew the auth ticket if KeepUserLoggedIn is true and configured to clear all appropriate cookie values on logout
* Fixes issue with a cookie that keeps growing, we were using the `Actor` property of the ClaimsIdentity which we shouldn't be doing. In netcore this is serialized whereas before it wasn't. And since we keep assigning it, each time we re-issued the auth ticket it would be another nested copy of the identity
* Fixes issue with GlobalSettings using the wrong key

## Testing

* Open developer tools and watch XHR requests
* Log in to the back office
* Wait a little bit and you will start seeing requests for GetRemainingTimeoutSeconds which will return the seconds remaining
* Logout, the request should be successful and you will see headers to remove all appropriate cookies
* enter the /umbraco url into the browser to which will redirect to the login page confirming you are logged out

## Testing keep user logged in

Add this config to appsettings.json:

```json
{
    "Umbraco": {
        "CMS": {
            "Global": {
                "TimeOutInMinutes":  2
            },
            "Security": {
                "KeepUserLoggedIn": true,

            }
     }
}
```

Then log into the back office with dev tools open and watch the GetRemainingTimeoutSeconds calls. You'll see that every few calls to this it will auto-renew the ticket, you can see this by the SET-COOKIE headers and you'll also notice the response value of this shrink on each request but then when it's renewed it will be back up around 90 seconds. The user will never be logged out while the browser is open because the cookie keeps on renewing.
